wallet/src/state.rs

```diff
@@ -188,12 +227,168 @@ impl ClientState {
         }
     }
 
-    /// Generate a new transaction.
-    ///
-    /// TODO: this function is too complicated, merge with
-    /// builder API ?
+    /// Generate a new transaction delegating stake
+    #[instrument(skip(self, rng))]
+    pub fn build_delegate<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        rate_data: RateData,
+        unbonded_amount: u64,
+        fee: u64,
+        source_address: Option<u64>,
+    ) -> Result<Transaction, anyhow::Error> {
+        // If the source address is set, send the delegation tokens to the same
+        // address; otherwise, send them to the default address.
+        let (_label, self_address) = self
+            .wallet()
+            .address_by_index(source_address.unwrap_or(0) as usize)?;
+
+        let mut tx_builder = Transaction::build_with_root(self.note_commitment_tree.root2());
+
+        tx_builder
+            .set_fee(fee)
+            // XXX populate chain id from params?
+            .set_chain_id(CURRENT_CHAIN_ID.to_string())
+            .add_delegation(&rate_data, unbonded_amount);
+
+        let spend_amount = unbonded_amount + fee;
+        let mut spent_amount = 0;
+
+        for note in self.notes_to_spend(rng, spend_amount, &*STAKING_TOKEN_DENOM, source_address)? {
+            spent_amount += note.amount();
+            tx_builder.add_spend(
+                rng,
+                &self.note_commitment_tree,
+                self.wallet.spend_key(),
+                note,
+            )?;
+        }
+
+        let delegation_note = tx_builder.add_output_producing_note(
+            rng,
+            &self_address,
+            Value {
+                amount: rate_data.delegation_amount(unbonded_amount),
+                asset_id: rate_data.identity_key.delegation_token().id(),
+            },
+            memo::MemoPlaintext([0u8; memo::MEMO_LEN_BYTES]),
+            self.wallet.outgoing_viewing_key(),
+        );
+
+        let change_amount = spent_amount - spend_amount;
+        // TODO: support dummy notes, and produce a change output unconditionally.
+        // let change_note = if change_amount > 0 { ... } else { /* dummy note */}
+        if change_amount > 0 {
```

Comment:
This will be important since we reveal the actions in a transaction, and we'd ideally not reveal whether or not the delegation spent an exact note amount, since that leaks a fair amount of informaiton.